api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2AuthenticationProvider.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.auth.oauth2;
+
+import java.util.Objects;
+import java.util.function.Function;
+import org.projectnessie.client.auth.NessieAuthenticationProvider;
+import org.projectnessie.client.http.HttpAuthentication;
+import org.projectnessie.client.http.HttpClient;
+
+public class OAuth2AuthenticationProvider implements NessieAuthenticationProvider {
+
+  public static final String AUTH_TYPE_VALUE = "OAUTH2";
+
+  @Override
+  public String getAuthTypeValue() {
+    return AUTH_TYPE_VALUE;
+  }
+
+  @Override
+  public HttpAuthentication build(Function<String, String> configSupplier) {
+    OAuth2ClientParams params = OAuth2ClientParams.fromConfig(configSupplier);
+    OAuth2Client client = new OAuth2Client(params);
+    client.start();
+    return create(client);
+  }
+
+  public static HttpAuthentication create(OAuth2Authenticator authenticator) {
+    return new OAuth2Authentication(authenticator);
+  }
+
+  private static class OAuth2Authentication implements HttpAuthentication {
+
+    private final OAuth2Authenticator authenticator;
+
+    private OAuth2Authentication(OAuth2Authenticator authenticator) {
+      Objects.requireNonNull(
+          authenticator,
+          "OAuth2Authenticator must not be null for authentication type " + AUTH_TYPE_VALUE);
+      this.authenticator = authenticator;
+    }
+
+    @Override
+    public void applyToHttpClient(HttpClient.Builder client) {
+      client.addRequestFilter(
+          ctx -> {
+            AccessToken token = authenticator.authenticate();
+            ctx.putHeader(
+                "Authorization", capitalize(token.getTokenType()) + " " + token.getPayload());
```

Comment:
Would it be possible to add a test case for this to `AbstractOpenIdAuthentication` (if token type can be something other than `Bearer`)?

Reply:
To be fair, I don't think it can be something else than `(B|b)earer`. I wonder if we shouldn't just remove this `capitalize` stuff and print `Bearer` all the time.

Reply:
I'd omit `capitalize()` completely.
For example, [RFC 6749 section 7.1](https://www.rfc-editor.org/rfc/rfc6749.html#section-7.1) mentions a `MAC` authentication scheme (work in process). Not saying we should think of that at this time.
But overall, it can't probably be anything else than `Bearer` - maybe we should add a runtime check like `tokenType.toLowerCase(ROOT).equals("bearer")`?

Reply:
I like your suggestion, let's remove the capitalize stuff and replace with a check.